src/NanopassSharp/PassTransformer.cs

```diff
@@ -11,8 +18,251 @@ public static class PassTransformer
     /// <param name="transformations">The transformations to apply.</param>
     /// <returns></returns>
     /// <exception cref="NotImplementedException"></exception>
-    public static AstNodeHierarchy ApplyTransformations(AstNodeHierarchy tree, PassTransformations transformations)
+    public static AstNodeHierarchy ApplyTransformations(AstNodeHierarchy tree, ITransformations transformations)
     {
-        throw new NotImplementedException();
+        if (transformations.Count == 0)
+        {
+            return tree;
+        }
+
+        tree = ApplyTransformationsToTree(tree, transformations);
+
+        var builder = AstNodeHierarchyBuilder.FromHierarchy(tree);
+
+        TraverseRoots(builder, tree, transformations);
+
+        return builder.Build();
+    }
+
+    /// <summary>
+    /// Traverses all the roots of a tree.
+    /// </summary>
+    /// <param name="builder">The builder which is being transformed.</param>
+    /// <param name="tree">The tree which is being transformed.</param>
+    /// <param name="transformations">The transformations to apply.</param>
+    private static void TraverseRoots(AstNodeHierarchyBuilder builder, AstNodeHierarchy tree, ITransformations transformations)
```

Comment:
The public API as a static function is fine, but internally, you are defining many many data clumps and carry on the tree, node and transformations through every call through many nested method calls. Why not make this a non-static class, make the ctor private and store transformation constants as private members? The public API method would instantiate the class, pass in stuff in the ctor, then you can work with a nonstatic context, easing on your signatures a bit.